### ISSUE TITLE
teleinfo component broken for Linky Standard mode

### ISSUE THREAD

**Issue by Jpsy** [esphome/issues#2060]
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Hass.io 2021.5.4 on RPi4b

**ESP (ESP32/ESP8266, Board/Sonoff):**

D1 mini
Also tried another ESP8266 based MCU.

**ESPHome version (latest production, beta, dev branch)**

latest production (1.17.2)

**Affected component:**

teleinfo
https://esphome.io/components/sensor/teleinfo.html

**Description of problem:**

The `teleinfo` component seems to be broken for Standard mode of the Linky meter.
Historical mode seems to work fine.
In Standard mode the component does not decode the data that is received from the Linky meter.

I can verify that the serial data is well received by my ESP node at 9600 baud (see data dump below). 

But the component fails to read the tags from the data stream. 
The entities for each defined tag are created in HA but their state remains `unknown` and the ESPHome log does not show any recognized tags. Instead it only shows occasional warnings of type `[W] [teleinfo: 058]: Internal buffer full`.

**Problem-relevant YAML-configuration entries:**

This is a basic config to reproduce the problem:
```yaml
esphome:
  name: teleinfo_standard_example
  platform: ESP8266
  board: d1_mini

wifi:
  ssid:     !secret my_wlan_sid
  password: !secret my_wlan_pwd

logger:
  level: VERBOSE
  baud_rate: 0

api:

ota:

uart:
  id: uart_bus
  tx_pin: GPIO1
  rx_pin: GPIO3
  baud_rate: 9600
  parity: EVEN
  data_bits: 7

sensor:
  - platform: teleinfo
    update_interval: 15s
    historical_mode: false
    tags:
      - tag_name: "EAST"
        sensor:
          name: "Index (Wh)"
          unit_of_measurement: "Wh"
          icon: mdi:home-analytics
```

**Logs (if applicable):**

Debug log:
The log shows only this message, which repeats every few seconds:
```
[17:23:53] [W] [teleinfo: 058]: Internal buffer full
```

**Additional information and things you've tried:**

I have verified that I receive correct data from Linky by using another very simple ESPHome config that only reads in the UART data line by line and dumps it to the log.
Here is a sample of the result:

```
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'DATE	E210515111058		<'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'NGTF	      BASE      	<'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'LTARF	      BASE      	F'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'EAST	020009875	.'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'EASF01	020009875	A'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'EASF02	000000000	#'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'EASF03	000000000	$'
[11:10:56][D][text_sensor:015]: 'uart_readline': Sending state 'EASF04	000000000	%'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASF05	000000000	&'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASF06	000000000	''
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASF07	000000000	('
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASF08	000000000	)'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASF09	000000000	*'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASF10	000000000	"'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASD01	020009875	?'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASD02	000000000	!'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASD03	000000000	"'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'EASD04	000000000	#'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'IRMS1	000	.'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'IRMS2	001	0'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'IRMS3	002	2'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'URMS1	235	D'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'URMS2	229	H'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'URMS3	228	H'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'PREF	30	B'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'PCOUP	30	\'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SINSTS	00858	['
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SINSTS1	00063	@'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SINSTS2	00239	F'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SINSTS3	00557	J'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN	E210515104312	05455	6'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN1	E210515081707	01635	/'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN2	E210515082127	01962	0'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN3	E210515104312	03343	#'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN-1	E210514053152	05328	W'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN1-1	E210514053152	01564	F'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN2-1	E210514125148	03493	O'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'SMAXSN3-1	E210514210703	02541	A'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'CCASN	E210515110000	02962	;'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'CCASN-1	E210515103000	00632	S'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'UMOY1	E210515111000	235	&'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'UMOY2	E210515111000	225	&'
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'UMOY3	E210515111000	234	''
[11:10:57][D][text_sensor:015]: 'uart_readline': Sending state 'STGE	003A0001	:'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'MSG1	PAS DE          MESSAGE         	<'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'PRM	055156........	3'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'RELAIS	000	B'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'NTARF	01	N'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'NJOURF	00	&'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'NJOURF+1	00	B'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'PJOURF+1	00008001 NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTIL'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'ADSC	0418........	A'
[11:10:58][D][text_sensor:015]: 'uart_readline': Sending state 'VTIC	02	J'
```

**Comment by Jpsy** [comment]
I found the source of the problem.
A three-phase Linky in Standard mode creates 53 different tags with a total data length of more than 1300 bytes (see dump below).
This is more than the reserved `MAX_BUF_SIZE` of 1024. Hence the `Internal buffer full` messages. 
Setting the buffer size to 2048 solves the problem.

I will create a PR soon.

Dump of a 3-phase Linky data stream in Standard mode (excluding the leading/trailing 0x2/0x3 control chars):

    ADSC	041876163178	A
    VTIC	02	J
    DATE	E210519075608		J
    NGTF	      BASE      	<
    LTARF	      BASE      	F
    EAST	020084915	,
    EASF01	020084915	?
    EASF02	000000000	#
    EASF03	000000000	$
    EASF04	000000000	%
    EASF05	000000000	&
    EASF06	000000000	'
    EASF07	000000000	(
    EASF08	000000000	)
    EASF09	000000000	*
    EASF10	000000000	"
    EASD01	020084915	=
    EASD02	000000000	!
    EASD03	000000000	"
    EASD04	000000000	#
    IRMS1	000	.
    IRMS2	001	0
    IRMS3	003	3
    URMS1	227	E
    URMS2	237	G
    URMS3	241	C
    PREF	30	B
    PCOUP	30	\
    SINSTS	00883	Y
    SINSTS1	00050	<
    SINSTS2	00211	<
    SINSTS3	00623	D
    SMAXSN	E210519065305	05215	<
    SMAXSN1	E210519065313	01482	.
    SMAXSN2	E210519071212	01734	*
    SMAXSN3	E210519065237	02042	.
    SMAXSN-1	E210518072525	07184	"
    SMAXSN1-1	E210518072423	01597	R
    SMAXSN2-1	E210518055139	01704	N
    SMAXSN3-1	E210518072528	04003	K
    CCASN	E210519073000	00762	C
    CCASN-1	E210519070000	01798	(
    UMOY1	E210519075000	233	1
    UMOY2	E210519075000	231	0
    UMOY3	E210519075000	238	8
    STGE	003A0001	:
    MSG1	PAS DE          MESSAGE         	<
    PRM	05515629430109	3
    RELAIS	000	B
    NTARF	01	N
    NJOURF	00	&
    NJOURF+1	00	B
    PJOURF+1	00008001 NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTIL

**Comment by bazarboobi** [comment]
Bonsoir,

Je me pose plusieurs questions très simple mais très bête il existe des modules : Module Téléinfo USB - Cartelectronic USBTICLC, il y a eu un temps un module : retrieve electricity use from Linky et les ondes du linky on peut pas les capter ?

J'ai vu qu'un plugin existe sous domoticz se basant sur enedis.

Peut être comme beaucoup de personne avec un Linky comment avoir les bonnes info et les bonne solution, sans vouloir toucher a cette boîte verte ?

Existe t'il des solutions utilisant le compte enedis ? 

Merci !

**Comment by Jpsy** [comment]
Ce composant pour ESPHome fonctionne uniquement avec un circuit supplémentaire qui doit être construit par vous-même et connecté au Linky.

**Comment by DrCoolzic** [comment]
Hi,
I have exactly the same conditions (ESP8266 + Linky three-phases ...) and obviously the same problem of "buffer full".
I tested in historical without problem but I ask to switch to standard mode and when changed this morning boom it was broken!

This is my first ESPHome project so I may ask stupid questions...
1) Seems like you have open a bug fix request? and that it is still open. Any idea how long it will take for the bug to be fixed in the release software?
2) is there a way to fix the problem locally? I mean is there a way to create a build locally with the new buffer size and load it to the ESP? Currently I just use the "install" command and magically everything compiles and loads. This is the only way I know.!

Can you help ?
Thanks - Jean

**Comment by DrCoolzic** [comment]
I have found a local copy  of teleinfo.h in config\esphome\linky-information\src\esphome\components\teleinfo
and I have modified the buffer size to 2048. Unfortunately when starting compilation the value is reverted back to 1024.
Is there a way to fix this problem locally because it does not seems that your fix is going to make it a near future ?
Thanks

**Comment by Jpsy** [comment]
To host a local version of the teleinfo integration you have to create this folder on your HA machine:
/config/esphome/custom_components/teleinfo/
Then copy all files and folders from the original integration into your new local folder (9 files, 2 folders):
https://github.com/esphome/esphome/tree/dev/esphome/components/teleinfo
Then edit teleinfo.h and change MAX_BUF_SIZE from 1024 to 2048.

Background: Any folders in /config/esphome/custom_components/ will override the component folders of ESPHome.

**Comment by DrCoolzic** [comment]
I have cloned esphome on my machine and I just tried your procedure but as soon as I start the compilation process all the .h and .cpp files are removed from the 
config\esphome\linky-information\src\esphome\custom_components\teleinfo directory? The only files that are not deleted are __init__.py files (3)

What am I doing wrong

**Comment by DrCoolzic** [comment]
Also I have looked at the code and I would be very surprise that the standard mode work for the tags with three values?
For example 
`SMAXSN	E210519065305	05215`
Did you test all the tags?

**Comment by DrCoolzic** [comment]
When I look at the data I receive there something very strange:
`ADSC	02.....	;
VTIC	02	J
DATE	E210718111041		9
NGTF	      BASE      	<
LTARF	      BASE      	F
EAST	028138222	+
EASF01	028138222	>
EASF02	000000000	#
EASF03	000000000	$
EASF04	000000000	%
EASF05	000000000	&
EASF06	000000000	'
EASF07	000000000	(
EASF08	000000000	)
EASF09	000000000	*
EASF10	000000000	"
EASD01	028138222	<
EASD02	000000000	!
EASD03	000000000	"
EASD04	000000000	#
IRMS1	002	0
IRMS2	002	1
IRMS3	002	2
URMS1	231	@
URMS2	231	A
URMS3	232	C
PREF	12	B
PCOUP	12	\
SINSTS	01407	R
SINSTS1	48370	M
SINSTS2	17742	M
SINSTS3	41187	N
SMAXSN	E210718001319	03680	<
SMAXSN1	E210718073628	00760	5
SMAXSN2	E210718110511	00640	"
SMAXSN3	E210718001320	03250	 
SMAXSN-1	E210717192509	03950	%
SMAXSN1-1	E210717094009	01010	C
SMAXSN2-1	E210717142752	02910	M
SMAXSN3-1	E210718000000	03020	3
CCASN	E210718110000	00986	D
CCASN-1	E210718103000	00986	$
UMOY1	E210718111000	231	'
UMOY2	E210718111000	231	(
UMOY3	E210718111000	233	+
STGE	003A0001	:
MSG1	     PAS DE          MESSAGE    	<
PRM	22.....	9
RELAIS	000	B
NTARF	01	N
NJOURF	00	&
NJOURF+1	00	B
PJOURF+1	00008001 NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE NONUTILE	9`

what is strange is 
`SINSTS	01407	R
SINSTS1	48370	M
SINSTS2	17742	M
SINSTS3	41187	N`
totally inconsistent ???
I went to the linky meter and on the display values are correct but during transmission the value reported do not make sense.
Could be a problem of reception but I doubt as the is consistently wrong and this would be the only place where it happen?
As the fields are checked for CRC I will see when the buffer size will be fixed

**Comment by Jpsy** [comment]
Please read my text exactly. It is all in there. Do NOT use src directories of ESPHome. They are not meant to be changed and you will loose your changes.

The tags with three values cannot be read at this time. I have another PR open for this. But it needs to be merged with the current master branch, which has other changes.

**Comment by DrCoolzic** [comment]
Thanks you I have been able to compile with the local version.
However it still does not work because I am getting garbage characters. 

May I ask you what circuitry you are using to demodulate the TIC info? I am using the schematic described everywhere (for example https://gammatroniques.fr/connaitre-sa-consommation-electrique-avec-home-assistant/) but the demodulator is not working at 9600 bds (see my pictures at https://www.casimages.com/u/jeanTransen/431815/ )

**Comment by Jpsy** [comment]
Here is the circuit that I drew for another PR of mine to improve the documentation:
https://deploy-preview-1197--esphome.netlify.app/components/sensor/teleinfo.html

Differences are:

- I pullup with 10k against 3.3 V, not 5 V. Pulling up against 5 V definitely risks to kill the ESP as its inputs are designed for 3.3 V only.
- I use RX not D7 for the input signal. Using D7 means that the ESP has to emulate the UART in software. RX provides a hardware UART.
- It is also very important that you use an optocoupler with an integrated reverse diode (i.e. the LTV814). If you have one without, add a red LED inverse to the internal LED in the opto.
- If this does not help:  Try to reverser I1 and I2. Try to replace the ESP - maybe it is already damaged by the 5V on its input.

**Comment by DrCoolzic** [comment]
I have used exactly the same schematic as yours with 1k for input, 10k for pullup & pulldown, LTV814, and BS170
I have modified my circuit to use 3.3v and connected Rx signal to pin GPIO3
I have modified the yaml to specify the new Rx and disable log
I have increased buffer length to 2048
Much better because now it reads characters but still not working
```
[16:10:19][C][api:096]:   Address: linky-tic.local:6053
[16:10:22][E][teleinfo:118]: No group found
[16:10:31][E][teleinfo:118]: No group found
[16:10:41][E][teleinfo:118]: No group found
[16:10:52][E][teleinfo:118]: No group found
[16:11:03][E][teleinfo:036]: bad crc: got 8 except 70
[16:11:03][E][teleinfo:036]: bad crc: got 60 except 58
[16:11:03][E][teleinfo:036]: bad crc: got 60 except 54
[16:11:03][E][teleinfo:036]: bad crc: got 70 except 64
[16:11:03][E][teleinfo:036]: bad crc: got 33 except 76
[16:11:03][E][teleinfo:036]: bad crc: got 33 except 89
[16:11:03][E][teleinfo:036]: bad crc: got 32 except 90
[16:11:03][E][teleinfo:036]: bad crc: got 50 except 69
[16:11:03][E][teleinfo:036]: bad crc: got 50 except 48
[16:11:03][E][teleinfo:036]: bad crc: got 69 except 67
[16:11:03][E][teleinfo:036]: bad crc: got 70 except 66
[16:11:03][E][teleinfo:036]: bad crc: got 71 except 67
[16:11:03][E][teleinfo:036]: bad crc: got 0 except 66
[16:11:03][E][teleinfo:036]: bad crc: got 80 except 57
[16:11:03][E][teleinfo:036]: bad crc: got 19 except 83
[16:11:03][E][teleinfo:036]: bad crc: got 16 except 76
[16:11:03][E][teleinfo:036]: bad crc: got 58 except 48
[16:11:03][E][teleinfo:036]: bad crc: got 32 except 55
[16:11:03][E][teleinfo:036]: bad crc: got 32 except 92
[16:11:03][E][teleinfo:036]: bad crc: got 90 except 86
[16:11:03][E][teleinfo:036]: bad crc: got 74 except 69
[16:11:03][E][teleinfo:036]: bad crc: got 74 except 72
[16:11:03][E][teleinfo:036]: bad crc: got 62 except 56
[16:11:03][E][teleinfo:036]: bad crc: got 70 except 66
[16:11:03][E][teleinfo:118]: No group found
[16:11:11][E][teleinfo:118]: No group found
```
...
![linky_std_newsch_1](https://user-images.githubusercontent.com/6078364/126173999-5b72de11-da90-4402-8ad7-a51c7c16fa46.PNG)
The problem is that at 9600 bit is about 100µs and the delay of demodulator at end of modulation is about 80µs which leave a very narrow bit probably  not decoded correctly by UART
Again I am stuck. In the LTV814 DS they indicates a rising time of 100µs on a 10k load and 20µs with a 1K load.
I wonder if it will work if I decrease the value of the pullup resistor to 4.7K ?
You do not have this problem of slow optocoupler ?
Thanks again and sorry to bother you

**Comment by Jpsy** [comment]
In fact I have another opto - just because it was lying around here.
One more problem that I had faced:
I had an Enedis Atome adapter connected at the same time to I1 / I2. This did not work at all.
In theory the Linky output should be well able to drive multiple devices. But I guess the Atome is just out of spec.

**Comment by DrCoolzic** [comment]
Ok I finally made it ! I have rebuild the demodulator from scratch on a breadboard to be able to test different values.
And now I have a nice square signal slightly shifter from the carier but about the same on falling and rising edges.
Therefore I totally destroyed my soldered version a rebuilt a new one to finally connect it to the ESP and .... it works.
Dont know what was the problem that was leading to this slowly rising falling signal?
The only thing I changed compared to your schematic is that the pulldown resistor connected to the opto is now 3.9K instead of 10k. This is suppose to get better timing but I will try 10K on the breadboard tomorow.

Yes normally many devices can be driven as long as they all obey the minimum resistor load. May be the Atom adapter is pouring tooo much current lowering the voltage for all devices?

Is there an approximate timeframe for the tags with 3 values?
I guess this will need to be specified by the user in the yaml file with an optional (for upward compatibility) parameter to indicate horodated tags?

Thanks again for your help.
![ltv814+bs170-Rin560-RpuFET=10k-RpdOpto=3 9k](https://user-images.githubusercontent.com/6078364/126205218-887b2051-27ff-464e-b749-92d9540bfc9c.PNG)

**Comment by Jpsy** [comment]
Glad to hear that you finally got it working.
All the details about how to use the timestamps are contained in the documentation PR that I have already linked above: https://esphome.io/components/sensor/teleinfo.html
I will see into it that I soon rebase the PRs so that they can be finally merged and become available to the public.

**Comment by DrCoolzic** [comment]
Great!  Very good solution. I did not read the complete page. 
I am not familiar with the ESPHome release process but I suppose the Teleinfo fixes will be pushed as part of an ESPHome release

I just tested all tags and everything seems to work as expected. The only thing strange is that the SINST1-SINST3 values are totally mad as I have explained above. I guess this is a problem with the Linky and obviously not the program ???

![Tic values](https://user-images.githubusercontent.com/6078364/126221838-1677fe30-18fb-4f88-b314-03901ab4b64c.PNG)
